[Security] Deprecate "AbstractVoter" in favor of "Voter"

lyrixx opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #16556, #16558, #16554 |
| License | MIT |
| Doc PR | - |

lyrixx committed:
[Security] Deprecate "AbstractVoter" in favor of "Voter"

stof commented on src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php:
```diff
@@ -53,10 +53,10 @@ public function supportsClass($class);
      *
      * @param TokenInterface $token      A TokenInterface instance
-     * @param object|null    $object     The object to secure
+     * @param mixed          $subject    The subject to secure
```
this is a BC break, as explained in #16600

stof commented:
I don't see any reason to deprecated AbstractVoter in favor of an abstract Voter class (which does not follow our naming conventions btw). This does not remove the BC break

lyrixx commented:
> I don't see any reason to deprecated AbstractVoter

I deprecated it because I added a new "base voter". 
I added a new "base voter" in order to fix listed issue.

lyrixx commented:
I reverted the change on the interface to ease integration of this PR.

nicolas-grekas commented:
:+1:
Status: reviewed

It removes a limitation of the current AbstractVoter (the first `if` block of the `vote` method) that can't be fixed with a simple deprecation notice. Which means we need a new class name. `Voter` is fine to me (abstract, like `Command` is).
@weaverryan I'd be happy to have your vote here since you cared about the linked issue.

weaverryan commented on UPGRADE-2.8.md:
````diff
@@ -475,6 +475,8 @@ Security
    ```
 
+ * The `AbstractVoter` is deprecated. Upgrade to `Voter` instead.
+
````
If we're deprecating `AbstractVoter` entirely, then we don't need to talk about it at all, right? The upgrade path would be to switch to `Voter` and make the changes needed there, correct? If so, I think we should have one simple, before and after (before `AbstractVoter` after with `Voter`).

lyrixx replied:
If people were using AbstractVoter in 2.7, they can read how to upgrade to 2.8;
Then they could upgrade with the new Voter. Like that it's done step by step.

But I can merge all steps together.

nicolas-grekas replied:
Or, we could and I think niw should juste revert the changes made on AbstractVoter for 2.8 entirely, but keep the deprecations. Ie remove the supports method & co. Don't you think ?

lyrixx replied:
This is a  real BC break

nicolas-grekas replied:
How that? SF 2.8 is not released, how can it be à BC break to revert the changes made on 2.8?

lyrixx replied:
Because the implementation sticks on the implementation of the 2.7 version.

weaverryan commented on src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php:
```diff
@@ -0,0 +1,85 @@
+    public function supportsAttribute($attribute)
+    {
+        throw new \BadMethodCallException('supportsAttribute method is deprecated since version 2.8, to be removed in 3.0');
+    }
```
If the `AbstractVoter` still works in 2.8 (but is deprecated of course), do we need these methods at all in the new class? When I switch to the new class, I will change to use the new methods.

lyrixx replied:
Yes, to **implements** the interface :/

weaverryan commented:
I left a few comments, but I'm "pro" this idea and I like the name.

weaverryan commented:
But I also think this can wait until 3.0... It's great, but I'm not sure it's important enough to stick into the release at the last second: there's too much else going on. If we do save this for 3.1, then I can update the `AbstractVoter` docs (I can't work on that til this gets a decision).

weaverryan commented:
Whoops - I meant wait until 3.1 - I updated my comment.

nicolas-grekas committed:
[Security] Revert changes made between 2.7 and 2.8-beta

nicolas-grekas commented:
We worked together with @lyrixx on this, the PR now has two commits:
- the first one reverts all changes made between 2.7 and 2.8-beta on AbstractVoter and deprecates it instead;
- the second one moves all this work made between 2.7 and 2.8-beta on the new `Voter` class.

The new `Voter` class is now more generic, at least it is generic enough to be reused by all the concrete voter we provide by default (to be done in an other PR).

We can postpone this deprecation to 3.1, but it would be quite unfortunate to ask devs to update their app twice. It's late but maybe not too late to fix this limitation now. That would reduce the burden significantly on this topic.


```diff
git diff symfony/2.7...d3c6d93dff9761cacc811976656574985fb37009 -- src/Symfony/Component/Security/Core/Authorization/Voter/
diff --git a/src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php b/src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php
index efa1562..5dcf787 100644
--- a/src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php
+++ b/src/Symfony/Component/Security/Core/Authorization/Voter/AbstractVoter.php
@@ -11,6 +11,8 @@
 
 namespace Symfony\Component\Security\Core\Authorization\Voter;
 
+@trigger_error('The '.__NAMESPACE__.'\AbstractVoter class is deprecated since version 2.8, to be removed in 3.0. Upgrade to Symfony\Component\Security\Core\Authorization\Voter\Voter instead.', E_USER_DEPRECATED);
+
 use Symfony\Component\Security\Core\User\UserInterface;
 use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
 
@@ -18,6 +20,8 @@ use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
  * Abstract Voter implementation that reduces boilerplate code required to create a custom Voter.
  *
  * @author Roman Marint<C5><A1>enko <inoryy@gmail.com>
+ *
+ * @deprecated since version 2.8, to be removed in 3.0. Upgrade to Symfony\Component\Security\Core\Authorization\Voter\Voter instead.
  */
 abstract class AbstractVoter implements VoterInterface
 {
diff --git a/src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php b/src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php
index d00ff1c..7e243f9 100644
--- a/src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php
+++ b/src/Symfony/Component/Security/Core/Authorization/Voter/VoterInterface.php
@@ -30,6 +30,8 @@ interface VoterInterface
      * @param string $attribute An attribute
      *
      * @return bool true if this Voter supports the attribute, false otherwise
+     *
+     * @deprecated since version 2.8, to be removed in 3.0.
      */
     public function supportsAttribute($attribute);
 
@@ -39,6 +41,8 @@ interface VoterInterface
      * @param string $class A class name
      *
      * @return bool true if this Voter can process the class
+     *
+     * @deprecated since version 2.8, to be removed in 3.0.
      */
     public function supportsClass($class);
 ```

stof commented on src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php:
```diff
@@ -0,0 +1,85 @@
+            $vote = self::ACCESS_DENIED;
+
+            if ($this->voteOnAttribute($attribute, $object, $token)) {
+                // grant access as soon as at least one voter returns a positive response
```
this should be ``at least one attribute``. We are inside the voter here. We don't have multiple voters

nicolas-grekas replied:
fixed

stof commented on src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    /**
+     * Determines if the attribute and object are supported by this voter.
```
``whether``, not ``if`` AFAICT

nicolas-grekas replied:
I kept `if`. I'd use whether if it were "attribute **or** object", but with "and", it sounds better to me. I a native wants to balance this, I'd be happy to have their feedback

stof commented on src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php:
```diff
@@ -0,0 +1,85 @@
+     * Determines if the attribute and object are supported by this voter.
+     *
+     * @param string $attribute An attribute
+     * @param mixed  $object    The object to secure
```
this should be renamed if it is not always an object

nicolas-grekas replied:
Well, grammatically, it's an object. But a synonym would be less confusing I agree. [Here](http://dico.isc.cnrs.fr/dico/tr/search_en?r=object&send=Translate) are some, any preference?

jvasseur replied:
I think subject could be used instead.

nicolas-grekas replied:
I just updated the PR to use "matter" instead, [as suggested by thesaurus.com(http://www.thesaurus.com/browse/object).

javiereguiluz replied:
`$matter` is not a common term and it may be difficult to understand it. Could we call it `$object` and add a note in the PHPdocs saying that this can also be a non-object? By the way, how common is to pass a non-object to this argument?

nicolas-grekas replied:
@javiereguiluz frequent enough to raise all this discussion and the related issues :)

weaverryan commented on src/Symfony/Component/Security/Core/Authorization/Voter/Voter.php:
```diff
@@ -0,0 +1,85 @@
+     *
+     * @return bool True if the attribute and matter are supported, false otherwise
+     */
+    abstract protected function supports($attribute, $matter);
```
`$matter` is a weird word in English - usually it only refers to the scientific meaning of matter (https://en.wikipedia.org/wiki/Matter).

I liked `$subject` better, and I think we should hint in the docs that this is often an object:

```php
/**
 * @param mixed $subject   The subject to secure (e.g. an object the user wants to access)
 */
```

nicolas-grekas replied:
Lets go for subject, thanks for the feedback. I used `The subject to secure, e.g. an object the user wants to access or any other PHP type`

nicolas-grekas commented:
Thank you all for the feedback, all comments are now resolved, votes pending :)

fabpot commented:
So, IIUC, the `AbstractVoter` class is the same as in 2.7 with the deprecation notices added and the new `Voter` class only removed the `if` statement that was problematic. Looks like a no-brainer to me. :+1:

weaverryan commented:
:+1: I agree - that bug was always something we wanted to fix - we should have just done it all at once (like this PR).

nicolas-grekas commented:
Thank you @lyrixx.
